others/freivalds_algorithm.cpp

```diff
@@ -0,0 +1,169 @@
+                     std::array<std::array<double, N>, N> C) {
+    // Printing matrix A
+    std::cout << "\nThe first matrix is:\n ";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << A[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+
+    // Printing matrix B
+    std::cout << "\nThe Second matrix is:\n ";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << B[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+
+    // Printing matrix C
+    std::cout << "The resultant matrix: \n";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << C[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+    // generate a random nx1 vector r containing 0/1 elements
+    std::array<std::array<const double, N>, 1> a{0};
+
+    // Multiply matrix B with vector 'r' which gives a nx1 vector 'br'
+    std::array<std::array<double, N>, 1> br = {};
+
+    for (int i = 0; i < N; i++) {
+        for (int k = 0; k < N; k++) {
+            br[i][0] = br[i][0] + B[i][k] * a[k][0];
+        }
+    }
+    // multiply matrix C with vector 'r' which gives a nx1 vector 'cr'
+    std::array<std::array<double, N>, 1> cr{0};
+    // double cr[N][1];
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < 1; j++) {
+            for (int k = 0; k < N; k++) {
+                cr[i][j] = cr[i][j] + C[i][k] * a[k][j];
+            }
+        }
+    }
+    // multiply matrix A with the nx1 matrix 'br' to form a nx1 vector 'abr'
+    std::array<std::array<double, N>, 1> abr{0};
+
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < 1; j++) {
+            for (int k = 0; k < N; k++) {
+                abr[i][j] = abr[i][j] + A[i][k] * br[k][j];
+            }
+        }
+    }
+    // compute vector p=abr - cr
+    for (int i = 0; i < N; i++) {
+        abr[i][0] -= cr[i][0];
+    }
+
+    // if vector p= 0 then A x B = C
+    // if vector p != 0 then A x B != C
+    bool flag = true;
+    for (int i = 0; i < N; i++) {
+        if (abr[i][0] == 0) {
+            continue;
+        } else {
+            flag = false;
+        }
+    }
+    if (flag == true) {
+        std::cout << "\nC is the product of matrix A and C\n";
+    } else {
+        std::cout << "\nC is not the product of matrix A and C\n";
+    }
+    return flag;
+}
+}  // namespace freivald
+}  // namespace others
+
+
+/**
+ * Test function 1 with A = [[1,0],[2,1]]
+ * B = [[5,4],[-5,1]]
+ * B = [[5,4],[5,9]]
+ * @returns none
+ */
+static void test_1() {
+    std::array<std::array<double, 2>, 2> A = {{{1, 0}, {2, 1}}};
+
+    std::array<std::array<double, 2>, 2> B = {{{5, 4}, {-5, 1}}};
+
+    std::array<std::array<double, 2>, 2> C = {{{5, 4}, {5, 9}}};
+
+    bool result = others::freivalds::CoppersmithAlgo(A, B, C);
+    std::cout << result << "\n";
+
+    assert(result == true);

```

Comment:
This test fails when I run it on my computer. There might be some non-determinism.